crates/paimon/src/spec/manifest_list.rs

```diff
@@ -16,15 +16,134 @@
+    }
+}
+
+/// This file includes several [`ManifestFileMeta`], representing all data of the whole table at the corresponding snapshot.
+pub struct ManifestListFactory {
+    file_io: FileIO,
+}
+
+/// The factory to read and write [`ManifestList`]
+impl ManifestListFactory {
+    pub fn new(file_io: FileIO) -> ManifestListFactory {
+        Self { file_io }
+    }
+
     /// Write several [`ManifestFileMeta`]s into a manifest list.
     ///
     /// NOTE: This method is atomic.
     pub fn write(&mut self, _metas: Vec<ManifestFileMeta>) -> &str {
         todo!()
     }
+
+    /// Read [`ManifestList`] from the manifest file.
+    pub async fn read(&self, path: &str) -> Result<ManifestList> {
+        let bs = self.file_io.new_input(path)?.read().await?;
+        // todo support other formats
+        ManifestList::from_avro_bytes(bs.as_ref())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::spec::ManifestList;
+
+    #[tokio::test]
+    async fn test_read_manifest_list() {
+        let workdir =
+            std::env::current_dir().unwrap_or_else(|err| panic!("current_dir must exist: {err}"));
+        let path = workdir
+            .join("tests/fixtures/manifest/manifest-list-5c7399a0-46ae-4a5e-9c13-3ab07212cdb6-0");
+        let v = std::fs::read(path.to_str().unwrap()).unwrap();
+        let res = ManifestList::from_avro_bytes(&v).unwrap();
+        assert_eq!(res.entries().len(), 2);
+        assert_eq!(
+            res.entries().first().unwrap().file_name(),
+            "manifest-0d7f646b-ae34-44ae-a60d-940aef93cbeb"
+        );
+        assert_eq!(res.entries().first().unwrap().version(), 2);
+        assert_eq!(res.entries().first().unwrap().file_size(), 700);
+        assert_eq!(res.entries().first().unwrap().num_added_files(), 7);
+        assert_eq!(res.entries().first().unwrap().num_deleted_files(), 0);
+        assert_eq!(res.entries().first().unwrap().schema_id(), 0);
+        // todo verify the max value, min value
+        assert!(!res
+            .entries()
+            .first()
+            .unwrap()
+            .partition_stats()
+            .max_values()
+            .is_empty());
+        assert!(!res
+            .entries()
+            .first()
+            .unwrap()
+            .partition_stats()
+            .min_values()
+            .is_empty());
+        assert!(!res
+            .entries()
+            .first()
+            .unwrap()
+            .partition_stats()
+            .null_counts()
+            .is_empty());
+
+        assert_eq!(
+            res.entries().get(1).unwrap().file_name(),
+            "manifest-88fe5442-3374-4385-a08f-022abd823a27"
+        );
+        assert_eq!(res.entries().get(1).unwrap().version(), 2);
+        assert_eq!(res.entries().get(1).unwrap().file_size(), 600);
+        assert_eq!(res.entries().get(1).unwrap().num_added_files(), 6);
+        assert_eq!(res.entries().get(1).unwrap().num_deleted_files(), 0);
```

Comment:
Use a file with a manifest-list that contains `num_deleted_files`?

Reply:
Sure, will update.

Reply:
Updated